src/net.cpp

```diff
@@ -1031,6 +1060,9 @@ void ThreadSocketHandler()
             BOOST_FOREACH(CNode* pnode, vNodesCopy)
                 pnode->Release();
         }
+
+        if (progress == 0)  // Nothing happened even though select did not block.  So slow us down.
+            MilliSleep(50);
```

Comment:
If rate limiting is off, data will always be read so "progress" will be
nonzero.  So this sleep will "never" occur if rate limiting is off.

"Never" is in quotes because the only time it would sleep is if there is
some kind of bug where nothing is read.  Presumably this means that the
looping call to select() will never block because the code fails to consume
any data.  In this case bitcoind would hit 100% CPU (if this sleep code was
not there).  So this code "helps" the non-shaping case by keeping the
system usable even if there is a bug in the select handling code.

Of course, any code that reduces the effects of bugs may also have the side
effect of hiding them... so you can collectively decide whether you want a
compound if statement, also checking whether the leaky bucket is on, here.

On Thu, Oct 15, 2015 at 9:32 AM, Peter Tschipper <notifications@github.com>
wrote:

> In src/net.cpp
> <https://github.com/bitcoinxt/bitcoinxt/pull/81#discussion_r42123885>:
>
> > @@ -1031,6 +1060,9 @@ void ThreadSocketHandler()
> >              BOOST_FOREACH(CNode* pnode, vNodesCopy)
> >                  pnode->Release();
> >          }
> > +
> > +        if (progress == 0)  // Nothing happened even though select did not block.  So slow us down.
> > +            MilliSleep(50);
>
> perhaps you could turn off the 50 millisecond sleep when rate limiting is
> not turned on so that it doesn't impact the performance for those wanting
> to rate limit...
> … <#1506bb3d7d8c6de1_>
> On 15/10/2015 6:23 AM, gandrewstone wrote: In src/net.cpp <#81 (comment)
> <https://github.com/bitcoinxt/bitcoinxt/pull/81#discussion_r42122953>>: >
> @@ -1031,6 +1060,9 @@ void ThreadSocketHandler() > BOOST_FOREACH(CNode*
> pnode, vNodesCopy) > pnode->Release(); > } > + > + if (progress == 0) //
> Nothing happened even though select did not block. So slow us down. > +
> MilliSleep(50); Select won't block because there is data available. But we
> are not "harvesting" it because there is not enough room in the leaky
> bucket. 50msec is unfortunately completely arbitrary. — Reply to this email
> directly or view it on GitHub <
> https://github.com/bitcoinxt/bitcoinxt/pull/81/files#r42122953>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoinxt/bitcoinxt/pull/81/files#r42123885>.
>